packages/framework/presence/src/presenceDatastoreManager.ts

```diff
@@ -0,0 +1,316 @@
+				{
+					[internalWorkspaceAddress]: {
+						[stateKey]: { [clientId]: value },
+					},
+				},
+				forceBroadcast,
+			);
+		};
+
+		const entry = createPresenceStates(
+			{
+				clientId: () => this.runtime.clientId,
+				lookupClient: this.presence.getAttendee.bind(this.presence),
+				localUpdate,
+			},
+			workspaceDatastore,
+			requestedContent,
+		);
+
+		this.workspaces.set(internalWorkspaceAddress, entry);
+		return entry.public;
+	}
+
+	private localUpdate(
+		data: GeneralDatastoreMessageContent & Partial<SystemDatastore>,
+		_forceBroadcast: boolean,
+	): void {
+		const content = {
+			sendTimestamp: Date.now(),
+			avgLatency: this.averageLatency,
+			// isComplete: false,
+			data,
+		} satisfies DatastoreUpdateMessage["content"];
+		this.runtime.submitSignal(datastoreUpdateMessageType, content);
+	}
+
+	private broadcastAllKnownState(): void {
+		this.runtime.submitSignal(datastoreUpdateMessageType, {
+			sendTimestamp: Date.now(),
+			avgLatency: this.averageLatency,
+			isComplete: true,
+			data: this.datastore,
+		} satisfies DatastoreUpdateMessage["content"]);
+		this.refreshBroadcastRequested = false;
+	}
+
+	public processSignal(
+		message: IInboundSignalMessage | DatastoreUpdateMessage | ClientJoinMessage,
+		local: boolean,
+	): void {
+		const received = Date.now();
+		assert(message.clientId !== null, "Map received signal without clientId");
+		if (!isPresenceMessage(message)) {
+			return;
+		}
+		if (local) {
+			const deliveryDelta = received - message.content.sendTimestamp;
+			this.returnedMessages = Math.min(this.returnedMessages + 1, 256);
+			this.averageLatency =
+				(this.averageLatency * (this.returnedMessages - 1) + deliveryDelta) /
+				this.returnedMessages;
+			return;
+		}
+
+		const timeModifier =
+			received -
+			(this.averageLatency + message.content.avgLatency + message.content.sendTimestamp);
+
+		if (message.type === joinMessageType) {
+			const updateProviders = message.content.updateProviders;
+			this.refreshBroadcastRequested = true;
+			// We must be connected to receive this message, so clientId should be defined.
+			// If it isn't then, not really a problem; just won't be in provider list.
+			// eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+			if (updateProviders.includes(this.runtime.clientId!)) {
+				// Send all current state to the new client
+				this.broadcastAllKnownState();
+			} else {
+				// Schedule a broadcast to the new client after a delay only to send if
+				// another broadcast hasn't been seen in the meantime. The delay is based
+				// on the position in the audience list. It doesn't have to be a stable
+				// list across all clients. We need something to provide suggested order
+				// to prevent a flood of broadcasts.
+				let indexOfSelf = 0;
+				for (const clientId of this.runtime.getAudience().getMembers().keys()) {
+					if (clientId === this.runtime.clientId) {
+						break;
+					}
+					indexOfSelf += 1;
+				}
```

Comment:
Consider using Array.findIndex() rather than custom logic.

Reply:
keys is not an array; just an iterable. But I already have a better mechanism in follow up change. Will only use quorum and its ordering.
```typescript
				const quorumMembers = this.runtime.getQuorum().getMembers();
				const indexOfSelf =
					quorumMembers.get(clientId)?.sequenceNumber ??
					// Index past quorum members + arbitrary additional offset up to 10
					quorumMembers.size + Math.random() * 10;
```